File: java/pgcon2/Q3.java
package pgcon2;

import java.io.BufferedReader;
import java.io.InputStreamReader;

public class Q3 {
	
	static StringBuffer table;
	
	public static void main(String[] args) throws Exception {
    	BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    	String line = br.readLine();
    	int len = line.length();
    	long max = 1 << (len-1);
    	int min = Integer.MAX_VALUE;
    	for (long i=0; i<max; i++) {
    		table = new StringBuffer();
    		for (int j=0; j<len; j++) {
    			long mask = 1 << j;
    			if ((i & mask) > 0) {
    				left(line.charAt(j));
    			} else {
    				right(line.charAt(j));
    			}
    		}
    		if (min > table.length()) {
    			min = table.length();
    		}
    	}
    	System.out.println(min);
	}
	
	static void left(char ch) {
		if (table.length() > 0) {
			if (table.charAt(0) == ch) {
				String s = table.substring(1);
				table = new StringBuffer();
				table.append(s);
			} else {
				table.insert(0, ch);
			}
		} else {
			table.append(ch);
		}
	}
	
	static void right(char ch) {
		if (table.length() > 0) {
			if (table.charAt(table.length()-1) == ch) {
				String s = table.substring(0, table.length()-1);
				table = new StringBuffer();
				table.append(s);
			} else {
				table.append(ch);
			}
		} else {
			table.append(ch);
		}
	}
	
}

File: c/pgcon2/q3.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_LEN ( 10000 )
#define _intMax ( 2147483647 )
typedef int boolean;
#define true  ( 1 )
#define false ( 0 )
#define max(a,b) ((a)>(b)?(a):(b))
#define min(a,b) ((a)<(b)?(a):(b))

char line[MAX_LEN];
#define MAX_SIZE ( 64 )
char elements[MAX_SIZE];
int head;
int tail;

void addFirst(char e) {
    head = (head - 1) & (MAX_SIZE - 1);
    elements[head] = e;
    if (head == tail) {
        printf("Deque is full");
    }
}

void addLast(char e) {
    elements[tail] = e;
    tail = (tail + 1) & (MAX_SIZE - 1);
    if (head == tail) {
        printf("Deque is full");
    }
}

char getFirst() {
    int h = head;
    char result = elements[h];
    return result;
}

char getLast() {
    int t = (tail - 1) & (MAX_SIZE - 1);
    char result = elements[t];
    return result;
}

char removeFirst() {
    int h = head;
    char result = elements[h];
    if (result == 0) return 0;
    elements[h] = 0;
    head = (h + 1) & (MAX_SIZE - 1);
    return result;
}

char removeLast() {
    int t = (tail - 1) & (MAX_SIZE - 1);
    char result = elements[t];
    if (result == 0) return 0;
    elements[t] = 0;
    tail = t;
    return result;
}

int size() {
    return (tail - head) & (MAX_SIZE - 1);
}

void left(char ch) {
    if (size() > 0) {
        if (getFirst() == ch) {
            removeFirst();
        } else {
            addFirst(ch);
        }
    } else {
        addFirst(ch);
    }
}

void right(char ch) {
    if (size() > 0) {
        if (getLast() == ch) {
            removeLast();
        } else {
            addLast(ch);
        }
    } else {
        addFirst(ch);
    }
}

int main() {
    int len;
	fgets(line, sizeof(line), stdin);
	len = strlen(line);
	if (line[len-1] == '\n') line[--len] = '\0';
	long max = 1L << (len-1);
	int min = _intMax;
	long i;
	int j;
	for (i=0; i<max; i++) {
	    head = 0;
	    tail = 0;
	    memset(elements, 0, sizeof(char)*MAX_SIZE);
	    for (j=0; j<len; j++) {
	        long mask = 1L << j;
	        if (i & mask) {
	            left(line[j]);
	        } else {
	            right(line[j]);
	        }
	    }
	    min = min(min, size());
	}
	printf("%d\n", min);
    return 0;
}

